mailbox/api/src/main/java/org/apache/james/mailbox/SessionProvider.java

```diff
@@ -42,6 +46,18 @@
      */
     MailboxSession createSystemSession(Username userName);
```

Comment:
Let's not take care of this now, but I am reluctant of having a `username` associated with a system session (it's a lie, and a non-sense).

Maybe we could have a `Tenant` API (empty) with a `UserTenant` and a `ServiceTenant` thus:
 - This would replace the `Session Type` + `Username` variable combinaison
 - Right resolution could be done with a pattern matching like syntax
 - And finally it would remove a user of `Username` class for something that **is not the name of an user** 

If done, it should be in a separate PR as it solves a connex problem.

Feedback welcomed on this proposal.

Reply:
Might need to discuss a bit more about details but in the idea I agree, a system session shouldn't depend on a regular username, it's non sense indeed.

Better in an other PR yes, this one is big enough already

Reply:
Sorry I don't get why it's a lie and a non-sense, can you explain me?

Reply:
Well a service do not look like much of a user to me, it's different context.

For instance, I don't want to care if my service is valid regarding `virtualHosting`, but I want to care about it for my usernames.  

Also, using a systemSession with a username makes think that you are doing impersonation, which is not enough to act on a messageId/mailboxId (as you do not have enougth information to know who you should impersonnate - root cause for this change-set). The system session is basicaly here trusted for any mailbox-api operation regardless of the 'username'.

Reply:
Thanks. So you should have 2 kinds of session? A user session carrying a username and a system session without it?

Reply:
Yes either 2 sub-interface of MailboxSession, but I suspect abstracting things away behind a tenant might simplify things...

Reply:
I just gave it a second or third thought:

With have a third use case in addition from "regular" user session, system service session: we have "delegated user" use case. Currently the "delegated user" part is fully ignored as it results in a 'standard user session'. Which from an audit perspective is terrible: if Bob have enabled delegation to Alice, and that a fraudulous action happened on Bob account, we are not able to identify who caused the action!

While I believe 3 MailboxSession sub APIs would be a big burdown for such a central API, working with *composition* with the use of a `Tenant` class could greatly help us!

My 2 cents...

Reply:
Could you explain what you mean by Tenant? I'm not sure how this word matches the concept we have here.